### PR TITLE
feat(batch-task): init task status management

### PR DESCRIPTION
**Pull request by BowenXiao1999**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
link: https://github.com/singularity-data/risingwave/issues/3834#issuecomment-1196306711

Considering that if introduce 2 rpcs, there will be a time delay between create_task and get_task_info, this may makes the   impl looks weird. So here I adopt the streaming RPC method. 

The create task return a stream, once it knows all task are running (task receiver inited), schedule next stage.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
ref: #3834

### CODE REVIEW

File: src/batch/src/task/task_manager.rs [hunk 1]
```diff
@@ -164,6 +164,14 @@ impl BatchManager {
             .ok_or(TaskNotFound)?
             .get_error())
     }
+
+    /// Return the receivers for streaming RPC.
+    pub fn get_task_receiver(
```
Comment (BowenXiao1999) [inline src/batch/src/task/task_manager.rs:169]: There is a hack: we will swap the state receiver and return as stream RPC result.

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 2]
```diff
@@ -330,19 +318,75 @@ impl StageRunner {
...
+            {
+                running_task_cnt += 1;
+                // The task running count should always less or equal than the registered tasks
+                // number.
+                assert!(running_task_cnt <= self.tasks.keys().len());
+                // All tasks in this stage have been scheduled. Notify query runner to schedule next
+                // stage.
+                if running_task_cnt == self.tasks.keys().len() {
+                    self.notify_schedule_next().await?;
+                }
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/distributed/stage.rs:352]: All tasks running, schedule next stage.

File: src/batch/src/task/task_execution.rs [hunk 3]
```diff
@@ -245,9 +252,20 @@ impl<C: BatchTaskContext> BatchTaskExecution<C> {
             .extend(receivers.into_iter().map(Some));
         let failure = self.failure.clone();
         let task_id = self.task_id.clone();
+
+        // After we init the output receivers, it's must safe to schedule next stage -- send
+        // TaskStatus::Running here.
```
Comment (fuyufjh) [inline src/batch/src/task/task_execution.rs:257]: By the way, please remove the unused states here:

https://github.com/singularity-data/risingwave/blob/946544c1200f7be02906537f2f2e1f23a3ca4e21/proto/task_service.proto#L20-L31

File: src/batch/src/task/task_execution.rs [hunk 4]
```diff
@@ -245,9 +252,20 @@ impl<C: BatchTaskContext> BatchTaskExecution<C> {
...
+        let (mut state_tx, state_rx) = tokio::sync::mpsc::channel(1024);
```
Comment (fuyufjh) [inline src/batch/src/task/task_execution.rs:258]: `2` is enough?

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 5]
```diff
@@ -330,19 +318,75 @@ impl StageRunner {
...
+
+                // TODO: handle task failure.
+            }
+            println!(
```
Comment (fuyufjh) [inline src/frontend/src/scheduler/distributed/stage.rs:356]:
```suggestion
            info!(
```

File: src/batch/src/task/task_execution.rs (hunk 3 above)
Comment (BowenXiao1999) [reply src/batch/src/task/task_execution.rs:257]: I will do a refactor later to clean out all unused task status

File: src/batch/src/task/task_execution.rs [hunk 6]
```diff
@@ -281,44 +304,70 @@ impl<C: BatchTaskContext> BatchTaskExecution<C> {
         Ok(())
     }
 
+    /// Change state and notify frontend for task status.
+    pub async fn change_state_notify(
```
Comment (liurenjie1024) [inline src/batch/src/task/task_execution.rs:308]: Is it possible to merge this method with `change_state`?

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM, just need to fix comments.

File: src/batch/src/task/task_execution.rs (hunk 6 above)
Comment (BowenXiao1999) [reply src/batch/src/task/task_execution.rs:308]: Yes I think it's ok.

But this reminds me another problem: local mode `.execute` do not return a status stream, so it do not have to send the status change. However, we are reuse a lot of code for create_task and execute, so in this PR, for `.execute` we still create a channel and send stuff into it, while the receiver will never consume it. It's a cost and may be refactored later..

Comment (liurenjie1024) [reply src/batch/src/task/task_execution.rs:308]: LGTM

**Review by liurenjie1024 — APPROVED** [review]

**Comment by BowenXiao1999** [comment]
> This PR causes deterministic e2e test to fail. I'm now debugging...🤪

Oops, let me help that. Do you mean this PR: https://github.com/singularity-data/risingwave/pull/4476 ?

**Comment by wangrunji0408** [comment]
Yes, thanks! But I guess it's madsim's problem.

**Comment by wangrunji0408** [comment]
A little clue: `CreateTask` reached compute node after `GetData`.

**Comment by BowenXiao1999** [comment]
I also got some clues: change_state_notify in `.async_execute()` should happend after we insert table id into the HashMap. But now it happens before that. 

It's not so easy to fix this according to current code structure.. Let me think...

**Comment by BowenXiao1999** [comment]
I have done a very quick verification and found that still not works (local sst): 
Here is the code: https://github.com/singularity-data/risingwave/tree/bw/fix-deterministic-found-bug
But my local test still fail. I switched back main branch and found the test still fail.  `./risedev sslt -- "e2e_test/batch/**/*.slt"` I doubt whether I test it correctly.

Will check the bug tmr.

**Comment by wangrunji0408** [comment]
It's correct and I can also reproduce the error.

**Pull request merged**